test/python/circuit/library/test_diagonal.py

```diff
@@ -37,8 +37,12 @@ class TestDiagonalGate(QiskitTestCase):
     )
     def test_diag_gate(self, phases):
         """Test correctness of diagonal decomposition."""
-        diag = [np.exp(1j * ph) for ph in phases]
-        qc = Diagonal(diag)
+        diag_dict = {}
+        for ph in phases:
+            i = np.random.randint(0, 16)
+            diag_dict[i] = np.exp(1j * ph)
+        diag = _dict_to_list(diag_dict)
+        qc = Diagonal(diag_dict)
         simulated_diag = Statevector(Operator(qc).data.diagonal()).data
         ref_diag = Statevector(diag).data
```

Comment:
It's typically not good to rely on private methods from the library in the test; it couples the test very strongly to the internals of how the function works, which makes it hard to change public behaviour without modifying the tests, even if the behaviour hasn't actually changed.  It's typically better to have the test very explicit _show_ what it's doing to make the reference, so that a reader of the test can verify just by reading the test that the behaviour being tested is correct.